### PR TITLE
Refactor TempDataOnDisk

### PR DESCRIPTION
**Pull request by vdimir**
### Changelog category (leave one):

- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
...

Requires     https://github.com/ClickHouse/ClickHouse/pull/66605    https://github.com/ClickHouse/ClickHouse/pull/66600

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit f2bacc820e7056545c3590d419d99e6708d27447 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/66606/f2bacc820e7056545c3590d419d99e6708d27447/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/66586/e8627601b52e28a4d4203cd365cd5e4693e2af57/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: src/Storages/MergeTree/MergeTask.cpp [hunk 1]
```diff
@@ -1703,7 +1620,7 @@ void MergeTask::ExecuteAndFinalizeHorizontalPart::createMergedStream() const
             sort_description,
             partition_key_columns,
             global_ctx->merging_params,
-            (is_vertical_merge ? RowsSourcesTemporaryFile::FILE_ID : ""),  /// rows_sources' temporary file is used only for vertical merge
+            (is_vertical_merge ? ctx->rows_sources_temporary_file : nullptr),  /// rows_sources temporaty file is used only for vertical merge
```
Comment (vdimir) [inline src/Storages/MergeTree/MergeTask.cpp:1623]:
```suggestion
            (is_vertical_merge ? ctx->rows_sources_temporary_file : nullptr),  /// rows_sources' temporary file is used only for vertical merge
```

**Comment by vdimir** [comment]
> Integration tests (aarch64) [5/6] — fail: 1, passed: 472 [Details](https://s3.amazonaws.com/clickhouse-test-reports/66606/60fa4be570f78b935d47f9cc6ed88fbd2b0b1440/integration_tests__aarch64__[5_6].html)
> Integration tests (asan, old analyzer) [6/6] — fail: 1, passed: 517 [Details](https://s3.amazonaws.com/clickhouse-test-reports/66606/60fa4be570f78b935d47f9cc6ed88fbd2b0b1440/integration_tests__asan__old_analyzer__[6_6].html)
> Integration tests (tsan) [6/6] — fail: 1, passed: 511 [Details](https://s3.amazonaws.com/clickhouse-test-reports/66606/60fa4be570f78b935d47f9cc6ed88fbd2b0b1440/integration_tests__tsan__[6_6].html)


https://github.com/ClickHouse/ClickHouse/issues/67604
https://github.com/ClickHouse/ClickHouse/issues/71237

**Comment by vdimir** [comment]
@davenger I've updated the code to revert the excess changes around MergeTask that we discussed, please take a look

**Review by davenger — CHANGES_REQUESTED** [review]

**Comment by vdimir** [comment]
Investigating failure https://s3.amazonaws.com/clickhouse-test-reports/66606/6917a37d5c978e20bbc4dbfea48c8460a3cea0da/stateless_tests__release_.html

File: src/Interpreters/Aggregator.cpp [hunk 2]
```diff
@@ -1639,11 +1642,23 @@ Block Aggregator::convertOneBucketToBlock(AggregatedDataVariants & variants, Are
     return block;
 }
 
+std::list<TemporaryBlockStreamHolder> & Aggregator::getTemporaryData()
+{
+    return tmp_files;
+}
```
[the remark this answers is not in the document]
Comment (vdimir) [reply src/Interpreters/Aggregator.cpp]: I suppose we can move ownership of `std::list<TemporaryBlockStreamHolder>` to `AggregatingTransform` in `initGenerate`

**Comment by vdimir** [comment]
One like and I'll merge this PR

**Review by davenger — APPROVED** [review]

**Pull request merged**